tensorpipe/common/cuda_loop.cc

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/common/cuda_loop.h>
+
+#include <tensorpipe/common/cuda.h>
+#include <tensorpipe/common/system.h>
+
+namespace tensorpipe {
+
+CudaLoop::CudaLoop() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  thread_ = std::thread([this, lock{std::move(lock)}]() mutable {
+    setThreadName("TP_CUDA_callback_loop");
+    for (;;) {
+      if (operations_.empty()) {
+        if (joined_ && pendingOperations_ == 0) {
+          break;
+        } else {
+          cv_.wait(lock);
+        }
+      }
+
+      std::deque<Operation> operations;
+      std::swap(operations, operations_);
+
+      lock.unlock();
+      for (auto& op : operations) {
+        op.callback(op.error);
+      }
+      lock.lock();
```

Comment:
Nit: there's probably no strict need to release the lock here, since I guess these callbacks will be produced through `eagerCallbackWrapper`, which defers to the loop, so there's no risk of the callback calling back into this class. But releasing the lock is probably better as it reduces contention, so this was just an FYI.